### PR TITLE
Enable stalled stream protection for uploads behind new behavior version

### PR DESCRIPTION
**Pull request by jdisanti**
This PR creates a new behavior version that enables stalled stream protection for upload streams by default.

----

_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

### CODE REVIEW

File: rust-runtime/aws-smithy-runtime/src/client/defaults.rs [hunk 1]
```diff
@@ -191,13 +192,13 @@ fn default_stalled_stream_protection_config_plugin_v2(
...
-                    .grace_period(Duration::from_secs(5))
-                    .build(),
-            );
+            let mut config =
+                StalledStreamProtectionConfig::enabled().grace_period(Duration::from_secs(5));
+            // Before v2024_03_28, upload streams did not have stalled stream protection by default
+            if !behavior_version.is_at_least(BehaviorVersion::v2024_03_28()) {
+                config = config.upload_enabled(false);
+            }
+            layer.store_put(config.build());
```
Comment (rcoh) [inline rust-runtime/aws-smithy-runtime/src/client/defaults.rs:201]: for debuggability, I wonder if we should alter the name of the plugin.

I would probably make two separate plugins that we select based on BMV instead, may be clearer and less likely to accidentally regress. That said, I don't have strong feelings.

File: aws/sdk/integration-tests/s3/tests/stalled-stream-protection.rs [hunk 2]
```diff
@@ -92,11 +92,10 @@ async fn test_stalled_stream_protection_defaults_for_upload() {
     let _ = tokio::spawn(server);
 
     let conf = Config::builder()
+        // Stalled stream protection MUST BE enabled by default. Do not configure it explicitly.
         .credentials_provider(Credentials::for_tests())
         .region(Region::new("us-east-1"))
         .endpoint_url(format!("http://{server_addr}"))
-        // TODO(https://github.com/smithy-lang/smithy-rs/issues/3510): make stalled stream protection enabled by default with BMV and remove this line
-        .stalled_stream_protection(StalledStreamProtectionConfig::enabled().build())
         .build();
```
Comment (rcoh) [inline aws/sdk/integration-tests/s3/tests/stalled-stream-protection.rs:99]: should we copy one of these tests, use the older BMV, and validate that stalled stream upload protection is _not_ enabled?

File: CHANGELOG.next.toml [hunk 3]
```diff
@@ -70,3 +70,15 @@ message = "Stalled stream protection on downloads will now only trigger if the u
 references = ["smithy-rs#3485"]
 meta = { "breaking" = false, "tada" = false, "bug" = true }
 author = "jdisanti"
+
+[[smithy-rs]]
+message = "Stalled stream protection on uploads is now enabled by default behind `BehaviorVersion::v2024_03_28()`. If you're using `BehaviorVersion::latest()`, you will get this change automatically by updating."
```
Comment (rcoh) [inline CHANGELOG.next.toml:75]:
```suggestion
message = "Stalled stream protection on uploads is now enabled by default behind `BehaviorVersion::v2024_03_28()`. If you're using `BehaviorVersion::latest()`, you will get this change automatically by running `cargo update`. Updating your SDK is not necessary, this change will happen when a new version of the client libraries are consumed."
```

**Review by rcoh — APPROVED** [review]
lgtm, couple of inline comments

File: rust-runtime/aws-smithy-runtime-api/src/client/behavior_version.rs [hunk 4]
```diff
@@ -3,17 +3,23 @@
...
 
-/// Behavior major-version of the client
+/// Behavior version of the client
 ///
 /// Over time, new best-practice behaviors are introduced. However, these behaviors might not be
 /// backwards compatible. For example, a change which introduces new default timeouts or a new
 /// retry-mode for all operations might be the ideal behavior but could break existing applications.
-#[derive(Clone)]
+#[derive(Copy, Clone, PartialEq)]
 pub struct BehaviorVersion {
```
Comment (aajtodd) [inline rust-runtime/aws-smithy-runtime-api/src/client/behavior_version.rs:14]: We should probably add a section to the developer guide on behavior versions, what they are and guidance on how to choose, and what changed in each.

Comment (jdisanti) [reply rust-runtime/aws-smithy-runtime-api/src/client/behavior_version.rs:14]: It's there - https://docs.aws.amazon.com/sdk-for-rust/latest/dg/client.html

**Pull request merged**